Boost: detect xframe options

amustaque97 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->

Fixes #21023

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* This PR is few changes are done in other PR link: https://github.com/Automattic/jetpack-boost-critical-css-gen/pull/20
* In the Jetpack repo I have added recommendations for `XFrameDenyError`

#### Jetpack product discussion NO
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use? NO
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Testing instructions are same as of the other PR https://github.com/Automattic/jetpack-boost-critical-css-gen/pull/20

#### Screenshot:
![Screenshot from 2021-09-10 23-21-13](https://user-images.githubusercontent.com/21127788/132897672-373f5dab-d388-440d-a8b6-79bf2cfb2cfc.png)

cc @thingalon

thingalon commented on projects/plugins/boost/app/assets/src/js/utils/describe-critical-css-recommendations.ts:
```diff
@@ -435,6 +435,33 @@ const errorTypeSpecs: { [ type: string ]: ErrorTypeSpec } = {
+	XFrameDenyError: {
+		describeSet: set =>
+			_n(
+				"Jetpack Boost couldn't proceed because of x-frame-deny configuration",
+				"Jetpack Boost couldn't proceed because of x-frame-deny configuration",
```
```suggestion
				"Jetpack Boost couldn't load the following page due to its X-Frame-Options configuration:",
				"Jetpack Boost couldn't load the following pages due to their X-Frame-Options configuration:",
```

thingalon commented on projects/plugins/boost/app/assets/src/js/utils/describe-critical-css-recommendations.ts:
```diff
@@ -435,6 +435,33 @@ const errorTypeSpecs: { [ type: string ]: ErrorTypeSpec } = {
+			),
+		rawError: set => Object.values( set.byUrl )[ 0 ].message,
+		suggestion: set => ( {
+			paragraph: _n(
+				'Please follow the troubleshooting steps below for the page.',
+				'Please follow the troubleshooting steps below for each of the pages.',
+				urlCount( set ),
+				'jetpack-boost'
+			),
```
The suggestion should give a user-friendly idea of what the error means.

```suggestion
			paragraph: __(
				'Jetpack Boost uses iframes while generating your Critical CSS. Unfortunately, your site has a special configuration header which prevents it from loading inside an iframe. The header is called "X-Frame-Options: DENY". This can be added to a WordPress site either by using a plugin, or by server configuration.',
				'jetpack-boost'
			),
```

thingalon commented on projects/plugins/boost/app/assets/src/js/utils/describe-critical-css-recommendations.ts:
```diff
@@ -435,6 +435,33 @@ const errorTypeSpecs: { [ type: string ]: ErrorTypeSpec } = {
+				'jetpack-boost'
+			),
+			list: [
+				__( 'Verify that x-frame-options header value is not set to deny', 'jetpack-boost' ),
+				__( 'If it does, <retry>try again</retry> to generate the Critical CSS.', 'jetpack-boost' ),
+				__(
+					'If the error still persist please contact <support>Jetpack Boost Support</support> with a copy of your error message.',
+					'jetpack-boost'
+				),
```
```suggestion
				__( 'Check that you are not using any plugins which add extra HTTP headers to your WordPress site, and deactivate them if you are.', 'jetpack-boost' ),
				__( 'If you are unsure of what these headers are, or where they come from please contact your hosting provider and ask them to remove the "X-Frame-Options" header from your site', 'jetpack-boost' ),
				__( '<retry>Try again</retry> to generate the Critical CSS.', 'jetpack-boost' ),
```

thingalon reviewed (changes requested):
Thank you for this fix; your implementation is good.

I've supplied some clearer wording for the error messages, and some suggestions which novice users will hopefully find easier to follow when facing this error. :)

amustaque97 commented:
Hi @thingalon, I have addressed the review comments, please take a look at PR once again. I'm attaching a screenshot after making changes.
![Screenshot 2021-09-13 21:41:21](https://user-images.githubusercontent.com/21127788/133119810-dace70c0-0e89-4795-a492-eb391bae61a2.png)

thingalon reviewed (approved):
👍  Looks great! Thank you for your contribution! :)

thingalon commented:
@amustaque97 would you mind merging or rebasing from `master`? Our pre-merge checklist won't let me merge if it's a little out of date. :)

amustaque97 commented:
@thingalon, I have taken a merge from master and PR is up to date. :)

thingalon reviewed (approved):
👍  Thanks!
